torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -4050,14 +4050,61 @@ Tensor linalg_matrix_exp_differential(
       self, grad, at::linalg_matrix_exp, /* adjoint */ adjoint);
 }
 
+template <typename F1, typename F2, typename... Ts>
+Tensor masked_fmap(
+    const Tensor& mask,
+    const F1& f1,
+    const F2& f2,
+    const Tensor& t,
+    const Ts&... ts) {
+  // This function takes two functions f1 and f2 and a (variadic) list of
+  // tensors, and creates a new tensor of the same shape as the first element of
+  // the list of tensors by applying the function f1 to the tensors for which
+  // the mask is true and f2 to the tensors for which the mask is false This
+  // function is used when we have a formula that works for, say, all
+  // non-singular inputs and another one for when the inputs are singular. See
+  // for example det_backward
+  auto n_true = mask.count_nonzero().item<int64_t>();
+  if (n_true == mask.numel()) {
+    return f1(t, ts...);
+  } else if (n_true == 0) {
+    return f2(t, ts...);
+  } else {
```

Comment:
Maybe we can skip these branches as to avoid syncs?

Reply:
It's how it was implemented before. I think it's fine.

Reply:
Yeah, but we can make it without any syncs whatsoever. Could you please provide some benchmarks to see whether the juice is wroth the squeeze?

Reply:
OK, I see that the previous implementation was syncing as well. Let's add a note/todo/comment about this?

Reply:
will do

Reply:
But I still believe that it is very likely that most use cases of `det` are full-rank matrices, maybe with a small subset of singular matrices :)

Reply:
That is true. And that's the reason for this sync, to completely avoid having to call the singular part at all.